src/bun.js/net.exports.js

```diff
@@ -368,7 +380,206 @@ export function createConnection(port, host, connectListener) {
+    this.#connectionListener = connectionListener;
+    this.#options = options;
+  }
+
+  ref() {
+    this.#server?.ref();
+    return this;
+  }
+
+  unref() {
+    this.#server?.unref();
+    return this;
+  }
+
+  close(callback) {
+    if (this.#server) {
+      this.#server.stop(true);
+      this.#server = null;
+      this.#listening = false;
+      this.#connections = 0;
+      this.emit("close");
+      if (typeof callback === "function") {
+        callback();
+      }
+
+      return this;
+    }
+
+    if (typeof callback === "function") {
+      callback(new Error("Server is not open"));
+    }
+    return this;
+  }
+
+  address() {
+    if (this.#server) {
+      //TODO: fix adress when host is passed
+      let address = this.#server.hostname;
+      const type = isIP(address);
+      const port = this.#server.port;
+      if (typeof port === "number") {
+        return {
+          port,
+          address,
+          family: type ? `IPv${type}` : undefined,
+        };
+      }
+
+      if (type) {
+        return {
+          address,
+          family: type ? `IPv${type}` : undefined,
+        };
+      }
+
+      return address;
+    }
+    return null;
+  }
+
+  getConnections(callback) {
+    if (typeof callback === "function") {
+      callback(this.#server ? null : new Error("Server is not open"), this.#connections);
+    }
+    return this;
+  }
+
+  listen(port, hostname, onListen) {
+    if (typeof hostname === "function") {
+      onListen = hostname;
+      hostname = undefined;
+    }
+
+    if (typeof port === "function") {
+      onListen = port;
+      port = 0;
+    } else if (typeof port === "object") {
+      port?.signal?.addEventListener("abort", () => this.close());
+
+      hostname = port?.host;
+      port = port?.port || 0;
+
+      // port <number>
+      // host <string>
+      // path <string> Will be ignored if port is specified. See Identifying paths for IPC connections.
+      // backlog <number> Common parameter of server.listen() functions.
+      // exclusive <boolean> Default: false
+      // readableAll <boolean> For IPC servers makes the pipe readable for all users. Default: false.
+      // writableAll <boolean> For IPC servers makes the pipe writable for all users. Default: false.
+      // ipv6Only <boolean> For TCP servers, setting ipv6Only to true will disable dual-stack support, i.e., binding to host :: won't make 0.0.0.0 be bound. Default: false.
+      // signal <AbortSignal> An AbortSignal that may be used to close a listening server.
+
+      if (typeof port?.callback === "function") onListen = port?.callback;
+    }
+    if (typeof port !== "number") {
+      port = 0;
+    }
+
+    hostname = hostname || "::";
+    const connectionListener = this.#connectionListener;
+    const { pauseOnConnect } = this.#options;
+    const self = this;
+    try {
+      this.#server = Bun.listen({
+        port,
+        hostname,
+        tls: false,
+        socket: {
```

Comment:
We shouldn't need to create this object in the server listener

Can we hoist it up to the root of the file and pass `data` to the socket to store the current instance of the `Server`?

Reply:
Sure I think will be a ServerHandlers in Socket symbol at this point

Reply:
Done on https://github.com/oven-sh/bun/pull/2337/commits/dc6bc3c6d8d73d3f5588116eeeaab409d01c765f